From 5056821d85179a6611f507fe28b6435eb5297d84 Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Mon, 26 Aug 2019 21:00:35 +0000
Subject: [PATCH] OS-7972 Early network admin service may not set the datalink
 MTU to the correct value Reviewed by: Dan McDonald <danmcd@joyent.com>
 Approved by: Brian Bennett <brian.bennett@joyent.com>

---
 usr/src/cmd/svc/milestone/net-early-admin | 26 +++++++++++++++++------
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/usr/src/cmd/svc/milestone/net-early-admin b/usr/src/cmd/svc/milestone/net-early-admin
index abfd5fbdd7..fbb2e4c5da 100644
--- a/usr/src/cmd/svc/milestone/net-early-admin
+++ b/usr/src/cmd/svc/milestone/net-early-admin
@@ -126,8 +126,8 @@ typeset -A mac_to_link
 out=$(dladm show-phys -mpo link,address)
 (( $? == 0 )) || fatal "dladm show-phys failed"
 while IFS=: read link addr; do
-    mac=$(normalize_mac $addr)
-    mac_to_link["$mac"]="$link"
+    macaddr=$(normalize_mac $addr)
+    mac_to_link["$macaddr"]="$link"
 done <<< "$out"
 
 ADMIN_NIC_TAG=${CONFIG_admin_tag:-"admin"}
@@ -147,11 +147,6 @@ if [[ -n "${aggr_links[$nic]}" ]]; then
 
     echo "Creating aggr: $nic (mode=$mode, links=${links})"
     dladm create-aggr -l ${links// / -l } -L $mode $nic
-
-    if [[ -n "${mtu[$nic]}" ]]; then
-        dladm set-linkprop -p mtu=${mtu[$nic]} $nic || \
-            fatal "ERROR: Failed to set mtu on aggr $nic to ${mtu[$nic]}"
-    fi
 elif valid_mac "$nic"; then
     [[ -n "${mac_to_link[$nic]}" ]] || \
         fatal "ERROR: admin mac address $nic not found on system"
@@ -160,6 +155,23 @@ else
     fatal "ERROR: Invalid value of ${ADMIN_NIC_TAG}_nic ($nic)"
 fi
 
+# If there are other nic tags configured on the same link as
+# the admin tag, find the largest MTU to use to set the
+# datalink MTU
+dlmtu="${mtu[$ADMIN_NIC_TAG]}"
+for tag in ${tags[@]}; do
+    tagmac="${tagv[$tag]}"
+    link="${mac_to_link[$tagmac]}"
+    if [[ "$link" == "$nic" && $dlmtu -lt ${mtu[$tag]} ]]; then
+        dlmtu=${mtu[$tag]}
+    fi
+done
+
+if [[ -n "$dlmtu" ]]; then
+    dladm set-linkprop -p mtu=$dlmtu $nic || \
+        fatal "ERROR: Failed to set mtu on aggr $nic to $dlmtu"
+fi
+
 driver=${nic%%[0-9]*}
 get_link_state $nic
 if [[ "$link_state" == "down" ]]; then
-- 
2.21.0

